### DISCUSSION TITLE
`TypedBlock` vs `Block`

### DISCUSSION THREAD

**Discussion by metasim** [FutureSDR/FutureSDR#153] [Q&A] [answered]
What are the pros/cons of implementing a new kernel block as a `TypedBlock` vs `Block` (especially given `Block::from_typed` exists)?

**Comment by bastibl** [comment, marked as answer]
Ideally, the `new` function of blocks should return a `TypedBlock` and `add_block` should accept `impl Into<Block>` https://github.com/FutureSDR/FutureSDR/blob/ae2ac61d9dad2917476bdf874ac65880f1f185c9/src/runtime/topology.rs#L119 but I didn't get around changing that yet.

The advantage of returning a `TypedBlock` is that you can call block (or rather `Kernel`) specific functions after the block is created. This is particularly helpful, for example, for an audio source, where you could query the sample rate of an audio file after it was opened by the block. If you have a block, you can also do that, but it is not that nice, since you'd first have to try to cast the kernel back to the given type.
